Add Directory.CreateTempSubdirectory

eerhardt opened the pull request:
CreateTempSubdirectory will create a new uniquely named directory under the temp directory. This allows applications to write temporary files in a co-located directory.

Contributes to #72881

Note: I've decided not to implement the full approved API in #72881 because of the feedback on the File API. That will be added in a future PR.

eerhardt committed:
Add Directory.CreateTempSubdirectory

CreateTempSubdirectory will create a new uniquely named directory under the temp directory. This allows applications to write temporary files in a co-located directory.

Contributes to #72881

jkotas commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs:
```diff
@@ -52,6 +53,29 @@ public static DirectoryInfo CreateDirectory(string path)
+
+        private static void EnsureNoDirectorySeparators(string? value, [CallerArgumentExpression("value")] string? paramName = null)
+        {
+            if (!string.IsNullOrEmpty(value) && value.IndexOfAny(s_directorySeparators) != -1)
```
```suggestion
            if (value.AsSpan().IndexOfAny(DirectorySeparators) < 0)
```
and change DirectorySeparators to constant.

eerhardt replied:
The issue is that on Windows there are 2 and on Unix there is only 1. 
To make it a constant, won’t I need to ifdef this code? What type would you expect the constant to be?

stephentoub replied:
It could be a string.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.Unix.cs:
```diff
@@ -20,5 +25,24 @@ private static DirectoryInfo CreateDirectoryCore(string path, UnixFileMode unixC
+            byte[] name = Encoding.UTF8.GetBytes(template);
+
+            // Create the temp directory.
+            byte* result = Interop.Sys.MkdTemp(name);
+            if (result == null)
```
This pointer is potentially pointing into garbage at this point, since it's the same pointer that was passed in to mkdtemp but the array is no longer pinned and could have been moved.  You're not actually using the pointer other than to see if it's null, so it's safe, but I'd suggest not even storing it into a local in order to discourage any such use, e.g.
```suggestion
            if (Interop.Sys.MkdTemp(name) is null)
```

jkotas replied:
Also, it would be nice for `MkdTemp` to take the unmanaged pointer as the argument and not depend on pinning via PInvoke marshalling for the same reason.

eerhardt replied:
Done. I also changed Path.GetTempFileName, since this code was cloned from there.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.Unix.cs:
```diff
@@ -20,5 +25,24 @@ private static DirectoryInfo CreateDirectoryCore(string path, UnixFileMode unixC
+        {
+            // mkdtemp takes a char* and overwrites the XXXXXX with six characters
+            // that'll result in a unique file name.
+            string template = $"{Path.GetTempPath()}{prefix}XXXXXX\0";
+            byte[] name = Encoding.UTF8.GetBytes(template);
```
Can we avoid this temporary byte[] allocation? e.g. something like:
```C#
string tempPath = Path.GetTempPath();
int length = tempPath.Length + (prefix?.Length ?? 0) + 6 + 1;
Span<byte> path = length <= 256 ? stackalloc byte[256] : new byte[length];
int pos = Encoding.UTF8.GetBytes(tempPath, path);
if (prefix is not null)
{
    pos += Encoding.UTF8.GetBytes(prefix, path.Slice(pos));
}
path.Slice(pos, 6).Fill((byte)'X');
path[pos + 6] = 0;
```

eerhardt replied:
Done. I also changed Path.GetTempFileName, since this code was cloned from there.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.Unix.cs:
```diff
@@ -20,5 +25,24 @@ private static DirectoryInfo CreateDirectoryCore(string path, UnixFileMode unixC
+            }
+
+            // 'name' is now the name of the directory
+            Debug.Assert(name[name.Length - 1] == '\0');
```
```suggestion
            Debug.Assert(name[^1] == '\0');
```

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.Windows.cs:
```diff
@@ -1,11 +1,68 @@
+            while (attempts < MaxAttempts)
+            {
+                Interop.GetRandomBytes(pKey, RandomKeyLength);
+                Path.Populate83FileNameFromRandomBytes(pKey, RandomKeyLength, builder.RawChars.Slice(builder.Length, RandomFileNameLength));
```
Why is it important to use the 8.3 support?

eerhardt replied:
It isn't important. But it is what GetRandomFileName uses

https://github.com/dotnet/runtime/blob/916f1ad5b391764e964b271f80b915869e1afed0/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs#L281-L289

I'm just simulating a call to `GetRandomFileName` here, without allocating an intermediate string.

eerhardt replied:
I can add a comment here, if that helps. Or did you have a different algorithm in mind for generating a folder name?

saul replied:
IMO having a period in the directory name seems a bit strange

eerhardt replied:
> IMO having a period in the directory name seems a bit strange

Does it really matter though? Code shouldn't depend on these names - they even have a different pattern between Windows and Unix. Besides, there are plenty of cases where directories have periods in their names - versions, namespace-like, etc.

e.g. `C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.7`.

Jozkee commented on src/libraries/System.IO.FileSystem/tests/Directory/CreateTempSubdirectory.cs:
```diff
@@ -0,0 +1,78 @@
+            {
+                Assert.True(tmpDir.Exists);
+                Assert.Equal(-1, tmpDir.FullName.IndexOfAny(Path.GetInvalidPathChars()));
+                Assert.Empty(Directory.GetFiles(tmpDir.FullName));
```
```suggestion
                Assert.Empty(Directory.GetFileSystemEntries(tmpDir.FullName));
```

Jozkee commented on src/libraries/System.IO.FileSystem/tests/Directory/CreateTempSubdirectory.cs:
```diff
@@ -0,0 +1,78 @@
+                if (!string.IsNullOrEmpty(prefix))
+                {
+                    Assert.StartsWith(prefix, tmpDir.Name);
+                }
```
Not very important but could we check that the prefix does not affect the unique name length? i.e: it keeps being 6 on Unix and 12 on Windows.

jozkee reviewed (approved):
Other than existing comments and a few more I'm leaving, LGTM.

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.Windows.cs:
```diff
@@ -1,11 +1,68 @@
+            // ensure the base TEMP directory exists
+            CreateDirectory(PathHelper.Normalize(ref builder));
+
+            builder.Append(prefix);
```
We also [doc](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.gettemppath) that this always returns a backslash, so I think we can safely assume there will always be a separator.

eerhardt commented on src/libraries/System.Runtime.Loader/tests/ResourceAssemblyLoadContext.cs:
```diff
@@ -42,10 +42,7 @@ protected override Assembly Load(AssemblyName assemblyName)
                 // This custom load context will extract that resource and store it at the %temp% path at runtime.
                 // This prevents the corerun from adding the test assembly to the TPA list.
                 // Once loaded it is not possible to unload the assembly, therefore it cannot be deleted.
-                var tempPath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
```
None of those places listed above are creating directories.

The first three are testing paths that don't exist. The last three are using files.

Jozkee replied:
Ah I didn't notice that, sorry.

eerhardt committed:
PR feedback

danmoseley commented on src/libraries/System.IO.FileSystem/tests/Directory/CreateTempSubdirectory.cs:
```diff
@@ -0,0 +1,80 @@
+        public void CreateTempSubdirectoryThrowsWithPrefixContainingDirectorySeparator(string prefix)
+        {
+            ArgumentException e = Assert.Throws<ArgumentException>(() => Directory.CreateTempSubdirectory(prefix));
+            Assert.Equal("prefix", e.ParamName);
```
AssertExtensions.Throws checks the param name.

eerhardt committed:
PR feedback

- Rewrite CreateTempSubdirectory and GetTempFileName on Unix to minimize allocations.

eerhardt committed:
Fix tests on Windows.

eerhardt commented:
I have responded to all existing feedback and pushed new changes. PTAL.

Nirmal4G commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs:
```diff
@@ -52,6 +53,29 @@ public static DirectoryInfo CreateDirectory(string path)
+        /// <returns>An object that represents the directory that was created.</returns>
+        /// <exception cref="ArgumentException"><paramref name="prefix" /> contains a directory separator.</exception>
+        /// <exception cref="IOException">A new directory cannot be created.</exception>
+        public static unsafe DirectoryInfo CreateTempSubdirectory(string? prefix = null)
```
```suggestion
        public static unsafe DirectoryInfo CreateTempSubDirectory(string? prefix = null)
```

Is it `Subdirectory` or `SubDirectory`?

eerhardt replied:
`Subdirectory` - see https://docs.microsoft.com/en-us/dotnet/api/system.io.directoryinfo.createsubdirectory

Nirmal4G replied:
Ahh... Even that should've been `SubDirectory` but too late it seems.

eerhardt replied:
https://www.merriam-webster.com/dictionary/subdirectory

says it is one word.

Nirmal4G replied:
<img width="400" alt="image" src="https://user-images.githubusercontent.com/4962580/183507074-66ead510-3524-468e-85fb-ddf4cf7bf41c.png">

It may be less common but in the programming world, especially in naming public APIs, the naming has to highlight individual words and acronyms. Of course, not every complex/compound word falls into this.

danmoseley replied:
I sen so not think there are many examples in core .NET API where we capitalize a letter in the middle of what is generally a single word. It would surprise me as a user.

tmds commented on src/libraries/System.IO.FileSystem/tests/Directory/CreateTempSubdirectory.cs:
```diff
@@ -0,0 +1,106 @@
+
+namespace System.IO.Tests
+{
+    public class Directory_CreateTempSubdirectory : FileSystemTest
```
Maybe verify that the filemode of the created directory is `700` on Unix. That is an important trait of this API.

tmds replied:
Q: is it on Windows also guaranteed the resulting directory is only accessible to the current user?

eerhardt replied:
> Maybe verify that the filemode of the created directory is 700 on Unix.

Done.

> Q: is it on Windows also guaranteed the resulting directory is only accessible to the current user?

No, on Windows there is no guarantee on the permissions of the directory. Just that it is empty, and it is in the TEMP directory.

tmds commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -9379,6 +9379,7 @@ public static partial class Directory
         [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("windows")]
         public static System.IO.DirectoryInfo CreateDirectory(string path, System.IO.UnixFileMode unixCreateMode) { throw null; }
         public static System.IO.FileSystemInfo CreateSymbolicLink(string path, string pathToTarget) { throw null; }
+        public static System.IO.DirectoryInfo CreateTempSubdirectory(string? prefix = null) { throw null; }
```
Was it actively decided to add `Sub` in this name? I prefer the slightly shorter: `CreateTempDirectory`.

Nirmal4G replied:
Me too but in the review, it was mentioned that it is actually a sub-directory within the temp folder. So...

eerhardt replied:
> Was it actively decided to add `Sub` in this name? I prefer the slightly shorter: `CreateTempDirectory`.

Yes, that was the decision in the [API review](https://youtu.be/nrpWy7kmZ28?t=2365). The reasoning is that it is more explicit. "Create Temp Directory" makes it sound like it is creating the actual Temp directory, and not a directory inside the Temp directory.

eerhardt committed:
Assert UnixFileMode is 700 for created temp directory.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.Unix.cs:
```diff
@@ -20,5 +23,37 @@ private static DirectoryInfo CreateDirectoryCore(string path, UnixFileMode unixC
+            int prefixByteCount = prefix is not null ? Encoding.UTF8.GetByteCount(prefix) : 0;
+            int totalByteCount = tempPathByteCount + prefixByteCount + 6 + 1;
+
+            Span<byte> path = totalByteCount <= 256 ? stackalloc byte[256].Slice(0, totalByteCount) : new byte[totalByteCount];
```
what is the benefit of stackalloc with const-size and slice over stackalloc with variable size?

```suggestion
            Span<byte> path = totalByteCount <= 256 ? stackalloc byte[totalByteCount) : new byte[totalByteCount];
```

eerhardt replied:
See the discussion here: https://github.com/dotnet/runtime/pull/63674#discussion_r783313413

> stackalloc of a constant size is more efficient than if you pass it a variable.
*provided you are compiling with [SkipLocalsInit], which runtime is.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/Directory.Unix.cs:
```diff
@@ -20,5 +23,37 @@ private static DirectoryInfo CreateDirectoryCore(string path, UnixFileMode unixC
+            {
+                if (Interop.Sys.MkdTemp(pPath) is null)
+                {
+                    Interop.CheckIo(-1);
```
This way of throwing exception for last error is unintuitive to me.

eerhardt replied:
I've added a new internal method `ThrowIOExceptionForLastError()` and used it in all places that were doing this.

adamsitnik reviewed (approved):
LGTM, thanks a lot @eerhardt !

eerhardt commented on src/libraries/Common/src/System/IO/TempFileCollection.cs:
```diff
@@ -150,6 +151,19 @@ private void EnsureTempNameCreated()
+        private string GetTempDirectory()
+        {
+            _createdTempDirectory = true;
+            return Directory.CreateTempSubdirectory().FullName;
```
This is actual BCL product code. It is used in two places:

1. (publicly exposed) in System.CodeDom
2. (internal) in System.Configuration.ConfigurationManager

Neither one of them use this temp directory directly. TempFileCollection actually doesn't even expose this location directly - instead it has a `BasePath` property that is set to an extension-less name for temp files created with this API:

https://github.com/dotnet/runtime/blob/e71a9583b4d6c9bd97edd87cda7f98f232f63530/src/libraries/Common/src/System/IO/TempFileCollection.cs#L111-L136

That BasePath is used to create additional files, with differing extensions:

https://github.com/dotnet/runtime/blob/e71a9583b4d6c9bd97edd87cda7f98f232f63530/src/libraries/Common/src/System/IO/TempFileCollection.cs#L66-L74

eerhardt commented on src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/Hosting/DirectoryCatalogDebuggerProxyTests.cs:
```diff
@@ -193,16 +191,12 @@ public static string GetRootTemporaryDirectory()
-            string tempDirectory = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
-            Directory.CreateDirectory(tempDirectory);
-            return tempDirectory;
+            return Directory.CreateTempSubdirectory().FullName;
```
That seems like a good follow-up PR. I can do that.

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/IO/Path.Unix.cs:
```diff
@@ -93,23 +93,34 @@ public static string GetTempPath()
+            fixed (byte* pPath = path)
+            {
+                IntPtr fd = Interop.CheckIo(Interop.Sys.MksTemps(pPath, SuffixByteLength));
+                Interop.Sys.Close(fd); // ignore any errors from close; nothing to do if cleanup isn't possible
```
> But it means that we should really consider introducing a new method 

Yep - that is the 2nd half of the approved API from https://github.com/dotnet/runtime/issues/72881#issuecomment-1203064309 (which I'm postponing for now and just adding the Directory piece for 7.0).

eerhardt committed:
PR feedback

eerhardt committed:
Rename ThrowExceptionForLastError to ThrowIOExceptionForLastError

adamsitnik reviewed (approved):
:shipit:
